### ISSUE TITLE
argparse can't handle inputs with spaces

### ISSUE THREAD

**Issue by TheThirdOne** [carykh/jumpcutter#23]
`./jump --input_file filename\ with\ spaces` will try to load from `filename`

Definitely seems like a problem in argparse, but there might be an easy way to make it still work.

Obviously there is an easy workaround of naming a file without spaces, but it would be nice to allow it.

**Comment by Bottersnike** [comment]
Does `--input_file "filename with spaces"` not work?

**Comment by nevercast** [comment]
Usually spaces are handled by the terminal, does your terminal usually work with escaped spaces?

**Comment by abhiTronix** [comment]
> Does --input_file "filename with spaces" not work?

@TheThirdOne Change "filename with spaces" to `filename\ with\ spaces`, that will solve your problem.

**Comment by balsoft** [comment]
I don't think it's an argparse problem - I believe it's because of `shell=True` in the ffmpeg invocations.

**Comment by balsoft** [comment]
@abhiTronix That's a hacky solution. We need to replace shell invocations with pure argument-list ones.

**Comment by balsoft** [comment]
This is actually the correct way to do it: https://github.com/balsoft/jumpcutter/blob/5a5adddd14cf4242cfb54ec26ae6a043f77b202c/jumpcutter.py#L116

**Comment by abhiTronix** [comment]
> @abhiTronix That's a hacky solution. We need to replace shell invocations with pure argument-list ones.
> 
> This is actually the correct way to do it: https://github.com/balsoft/jumpcutter/blob/5a5adddd14cf4242cfb54ec26ae6a043f77b202c/jumpcutter.py#L116
> 

@balsoft No its a problem with Unix systems, as they fail to take filename/ foldername with space(s) as input on the terminal. Here's the more insight into this problem: https://askubuntu.com/questions/516772/how-to-access-files-directories-with-spaces-in-the-name , Your solution will not work on the terminal, and the algorithm will fail before it even starts. Here is an example, try this:
`python jumpcutter.py --input_file abc/filename with space.mp4`

Hope that clears your doubt.

**Comment by TheThirdOne** [comment]
I was wrong with my initial diagnosis. I now agree with @balsoft, its a problem with the shell invocation which can be corrected by using argument based calls.

@abhiTronix, bash allows spaces in filenames passed as arguments, you just need to use quotes or escape the space with a slash. My original example shows escaping spaces as does the answer to the question you linked.

Ultimately I don't want something like
 `python jumpcutter.py --input_file abc/filename with space.mp4`
 to work, but I want
 `python jumpcutter.py --input_file abc/filename\ with\ space.mp4` 
  or
  `python jumpcutter.py --input_file "abc/filename with space.mp4"`
 to work.

**Comment by abhiTronix** [comment]
@TheThirdOne My answer is pointing out the workaround, I don't think the algorithm itself is responsible or changing the code will affect this, it is related to Terminal problem.  It will always be going to fail with error `no such file or directory`. Unless If we skip the part where we input "filename with space" file to the jumpcutter.py by --input_file flag and find a way to input file location directly to the algorithm without using the terminal and **_then fix it by using (" ") quotations as they can handle special characters such as spaces._**

**Comment by abhiTronix** [comment]
GUI Feature has been already been implemented in this [PR](https://github.com/carykh/jumpcutter/pull/35#issue-269602661) and will fix this issue.

**Comment by TheThirdOne** [comment]
@abhiTronix 
> It will always be going to fail with error no such file or directory without escaped spaces.

Absolutely, but I am using escaped spaces.  

A modification like what @balsoft made does in fact work. I have a working version on my computer (not their exact fork though).

Here is an example of the command I used an the output showing it reached the end of the script. 
```
$ python3 jumpcutter.py --input_file "sample file.mp4" --output_file out2.mp4 --silent_speed 999999 --sounded_speed 1
...
[libx264 @ 0x23bf7a0] i8 v,h,dc,ddl,ddr,vr,hd,vl,hu: 32% 21% 45%  1%  0%  0%  0%  0%  1%
[libx264 @ 0x23bf7a0] i4 v,h,dc,ddl,ddr,vr,hd,vl,hu: 35% 31% 14%  3%  3%  3%  4%  3%  5%
[libx264 @ 0x23bf7a0] i8c dc,h,v,p: 77% 17%  6%  0%
[libx264 @ 0x23bf7a0] Weighted P-Frames: Y:0.0% UV:0.0%
[libx264 @ 0x23bf7a0] ref P L0: 83.8%  1.8%  9.2%  5.1%
[libx264 @ 0x23bf7a0] ref B L0: 54.4% 41.0%  4.6%
[libx264 @ 0x23bf7a0] ref B L1: 88.5% 11.5%
[libx264 @ 0x23bf7a0] kb/s:515.28
[libvo_aacenc @ 0x23c06e0] 1 frames left in the queue on closing
```

It is absolutely possible to fix this, and the issue is the result of using
`subprocess.call(command, shell=true)`.

**Comment by balsoft** [comment]
@abhiTronix Yep, you're not escaping the filename - it'll ofc fail in this form. Everyone in the UNIX world knows about shell escape sequences, that allow to pass arguments that contain spaces. The problem was with shell invocation from Python, because python removed the escape sequences from input filename.

**Comment by balsoft** [comment]
> Does `--input_file "filename with spaces"` not work?

This is identical to OP's command.

**Comment by balsoft** [comment]
@TheThirdOne good.

**Comment by balsoft** [comment]
@abhiTronix GUI won't help -- it'll pass the same arguments to jumpcutter, and if it does say `["jumpcutter", "--input_file", "test file.mp4"`], that won't work. The problem is entirely in jumpcutter's script.

**Comment by abhiTronix** [comment]
> Here is an example of the command I used an the output showing it reached the end of the script.
> 
> $ python3 jumpcutter.py --input_file "sample file.mp4" --output_file out2.mp4 --silent_speed 999999 --sounded_speed 1
> ...
> [libx264 @ 0x23bf7a0] i8 v,h,dc,ddl,ddr,vr,hd,vl,hu: 32% 21% 45%  1%  0%  0%  0%  0%  1%
> [libx264 @ 0x23bf7a0] i4 v,h,dc,ddl,ddr,vr,hd,vl,hu: 35% 31% 14%  3%  3%  3%  4%  3%  5%
> [libx264 @ 0x23bf7a0] i8c dc,h,v,p: 77% 17%  6%  0%
> [libx264 @ 0x23bf7a0] Weighted P-Frames: Y:0.0% UV:0.0%
> [libx264 @ 0x23bf7a0] ref P L0: 83.8%  1.8%  9.2%  5.1%
> [libx264 @ 0x23bf7a0] ref B L0: 54.4% 41.0%  4.6%
> [libx264 @ 0x23bf7a0] ref B L1: 88.5% 11.5%
> [libx264 @ 0x23bf7a0] kb/s:515.28
> [libvo_aacenc @ 0x23c06e0] 1 frames left in the queue on closing

@TheThirdOne good, you used "" quotations in `--input_file` argument, that's another workaround for special characters like _spaces_ beside _escaped space_.

> GUI won't help -- it'll pass the same arguments to jumpcutter, and if it does say ["jumpcutter", "--input_file", "test file.mp4"], that won't work.
No, it can absolutely work, you just need to handle those string properly with f strings. Here's the Repo. owners [comment](https://github.com/carykh/jumpcutter/pull/9#issuecomment-482023133)

> The problem was with shell invocation from Python, because python removed the escape sequences from input filename.

@balsoft No. it doesn't.  Can you please elaborate on how python does that?

 > By no coincidence at all shell=True is called dangerous in official python documentation.
I agree with this statement.

**Comment by balsoft** [comment]
> No, it can absolutely work, you just need to handle those string properly with f strings. Here's the Repo. owner's comment

I was talking about the way things are implemented currently. It won't work in the current form.

> Can you please elaborate on how python does that?

I was oversimplifying, of course. The gist is this: a POSIX-compatible shell, from which the python process is executed, automatically parses the escape sequences and removes them from process arguments. E.g. if you do `echo "Hello"` you won't see the quotes around `Hello`, because the shell removed them. This means that whatever argparse returns as `input_file` won't contain quotes either - so it'll be `fiename with spaces`, not `"filename with spaces"`. This returned value then gets simply added to a ffmpeg command -- so it becomes `subprocess.call("ffmpeg -foo filename with spaces", shell=True)`, and then this shell which should invoke ffmpeg now sees three different arguments (filename, with, spaces) instead of one (`filename with spaces`). This is why your hack works - according to POSIX, shell shouldn't parse any escape sequences inside quotes, so the python is invoked with `filename'\ with'\ spaces` which then means that ffmpeg is invoked with `"ffmpeg -foo filename\ with\ spaces"`, so now the second shell parses the \ escape sequence three times and sees one argument (`filename with spaces`). So I was sort of wrong - it's not python that removes the escape sequences, it's a shell that calls python. What I suggest is that we ditch the second shell (which calls ffmpeg) because it's practically unneeded and specify arguments ourselves with `subprocess.call(["ffmpeg", "-foo", inputFileName])` so that ffmpeg sees exactly the argument that was passed to python. This fixes the issue.

>good, you used "" quotations in --input_file argument, that's another workaround for special characters like spaces beside escaped space.

In your answer, you've used both "workarounds" and because of that the second shell handled the input correctly, whereas for a well-behaved program just one of them should work. So either `jumpcutter --input_file filename\ with\ spaces.mp4` or `jumpcutter --input_file "filename with spaces.mp4"` must work. In current form it doesn't, with my solution it does.
